aapcs64-morello/aapcs64-morello.rst

```diff
@@ -517,6 +517,16 @@ The differences in language bindings used for AAPCS64 and AAPCS64-cap are descri
     |                              |                                                                                          |
     | A.3                          |                                                                                          |
     +------------------------------+------------------------------------------------------------------------------------------+
+    |                              | If the callee is variadic in AAPCS64-cap, the Anonymous Arguments Memory area is         |
+    |                              | allocated in memory with a size of at least 16 * (number of Anonymous arguments) bytes.  |
+    |                              | The Anonymous Arguments Memory area is 16-byte aligned. The Anonymous Arguments          |
+    |                              | Capability points to the Anonymous Arguments Memory area and is copied to C11. If there  |
+    | A.4                          | are no Anonymous arguments passed, C11 is set to NULL.                                   |
```

Comment:
Out of interest, why C11?

Also, this specifies that it must be nulled out on every non-variadic function call, right? That shouldn't add too much overhead because you can assume it's null inside a non-variadic function and thus only need to null it out if you allocate it yourself, but a function call could clobber it with unknown data so if you call `foo()` then `bar()` you can't guarantee that it's null any more for the second call. I think if you want this to be efficient you either need to make it only apply to variadic functions, or require that C11 also be null on return, that way it just becomes "if you allocate it you must clear it" and so can make it a higher-cost register to allocate.

(Note that prototype mismatches are already generally uncaught for CHERI, so I'm not sure if the specific case of variadic vs non-variadic needs special catching as it's just that problem; the wrong number/type of variadic arguments is however different because even if your prototypes match you can't check them without knowledge of the callee's semantics)

Reply:
C11 is the first register that's not used for anything else (c9 and c10 are used for cheri_ccall).

It's not nulled out on non-variadic function calls at the moment, but might be a good idea.

Reply:
FYI cheri_ccall(ee) are going to be removed because they're not good interfaces and the only user, the MIPS-specific libcheri, is already gone from CheriBSD. I wouldn't worry about them, just use C9. We never added support for it to CHERI-RISC-V.

Interestingly glibc does save/restore x9 to get a whole number of register pairs, which seems like a waste, but maybe there's a good reason for that. FreeBSD is even weirder, it saves xzr and "restores" it...

I wouldn't bother with it, as I say that's really just a special case of prototypes not matching the definition. I had missed that this whole step was guarded by "If the callee is variadic".